cmd/zpool/zpool_main.c

```diff
@@ -6909,6 +6909,13 @@ zpool_do_online(int argc, char **argv)
 					(void) printf(gettext("use 'zpool "
 					    "replace' to replace devices "
 					    "that are no longer present\n"));
+				if (flags & ZFS_ONLINE_EXPAND) {
```

Comment:
It seems unfortunate that this has to assume the details of the kernel's behavior.  It would be better if the code that's deciding to not expand would return an error.  But that would probably be difficult since there are a lot of layers between this code and where the logic is implemented in vdev_open():
```
	if (vd->vdev_state == VDEV_STATE_HEALTHY &&
	    ((asize > vd->vdev_asize &&
	    (vd->vdev_expanding || spa->spa_autoexpand)) ||
	    (asize < vd->vdev_asize)))
		vd->vdev_asize = asize;
```

It looks like it would be more in keeping with the surrounding logic if we printed a warning but still returned 0 (like the above `warning: onlined but remains in faulted state`).  There is some logic to the fact that `zpool online -e` is successful in other cases that don't actually expand the disks (e.g. if the disk is not actually bigger).

Another concern with the proposal is that it says we `failed to expand`, but we may have changed the partition size (the `wholedisk` case in `zpool_vdev_online()`).  Maybe that could be addressed with some additional wording in the error/warning message?

I wonder if it would be at all reasonable to expand the disk even if it is not healthy??? That would sidestep this problem entirely, but I'm not sure if there's a real reason that we didn't do that to begin with.

Reply:
Yeah I was hesitant to change that behavior (expand unhealthy disks) since I didn't understand the original restriction. That might be worth looking into; if we can just skip past this to get to the source by removing that restriction, that would certainly be simpler.

I originally was planning to have this change return an error in `vdev_open()` in that case, but the problem is that then 1) the reopen would fail, which it currently isn't allowed to do, and 2) importing pools with unhealthy expandable disks and autoexpand would fail, which would be a nontrivial change in behavior.

Perhaps a better version of this patch would be "If the vdev is degraded after the online && the expand flag was passed && the vdev was online before we started, then we fail"; that way we more closely capture the PR's intention of "if you only called `online -e` to expand the vdev and it didn't, that should fail" ?

I can certainly rephrase the message to make it clear that the partition size might have expanded even if the usable space didn't, though I do also wonder if we want to have consistent behavior between partition size changes and actual expansion (either both work on degraded vdevs or neither do) ?

Reply:
I also wonder about the possibility of just going ahead and expanding "degraded" drives. Logically, it seems like it should be a reasonable thing to do.